src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs

```diff
@@ -530,11 +646,49 @@ public override ConstructorInfo[] GetConstructors(BindingFlags bindingAttr)
         public override EventInfo[] GetEvents(BindingFlags bindingAttr) => throw new NotSupportedException();
         [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicEvents | DynamicallyAccessedMemberTypes.NonPublicEvents)]
         public override EventInfo? GetEvent(string name, BindingFlags bindingAttr) => throw new NotSupportedException();
+
         [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods | DynamicallyAccessedMemberTypes.NonPublicMethods)]
-        public override MethodInfo[] GetMethods(BindingFlags bindingAttr) => throw new NotSupportedException();
+        public override MethodInfo[] GetMethods(BindingFlags bindingAttr)
+        {
+            ThrowIfNotCreated();
+
+            List<MethodInfo> methods = new();
+            foreach (MethodBuilderImpl method in _methodDefinitions)
+            {
+                if (!method.IsConstructor && MatchesTheFilter(method, GetBindingFlags(method), bindingAttr, CallingConventions.Any, method.ParameterTypes))
+                {
+                    methods.Add(method);
+                }
+            }
+
+            return methods.ToArray();
+        }
+
         [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods | DynamicallyAccessedMemberTypes.NonPublicMethods)]
         protected override MethodInfo? GetMethodImpl(string name, BindingFlags bindingAttr, Binder? binder,
-                CallingConventions callConvention, Type[]? types, ParameterModifier[]? modifiers) => throw new NotSupportedException();
+                CallingConventions callConvention, Type[]? types, ParameterModifier[]? modifiers)
+        {
+            ThrowIfNotCreated();
+            ArgumentNullException.ThrowIfNull(name);
+
+            if (types != null)
+            {
+                for (int i = 0; i < types.Length; i++)
+                {
+                    ArgumentNullException.ThrowIfNull(types[i], nameof(types));
+                }
+            }
+
```

Comment:
```suggestion
```
This validation is also performed by the public entrypoint: https://github.com/dotnet/runtime/blob/f5d53478febc328c42e4efcb43d0646d9be176cd/src/libraries/System.Private.CoreLib/src/System/Type.cs#L329-L332